File: java/2.add-two-numbers.java
/*
 * @lc app=leetcode id=2 lang=java
 *
 * [2] Add Two Numbers
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode head = new ListNode();
        ListNode tail = head;

        int carry = 0;
        int sum;

        while(l1 != null || l2 != null || carry != 0) {
            if(tail.next != null) {
                tail = tail.next;
            }

            sum = 0;
            if(l1 != null) {
                sum += l1.val;
                l1 = l1.next;
            }

            if(l2 != null) {
                sum += l2.val;
                l2 = l2.next;
            }

            sum += carry;

            carry = sum / 10;
            
            sum %= 10;

            tail.val = sum;
            tail.next = new ListNode();
        }

        tail.next = null;

        return head;
    }
}
// @lc code=end

File: cpp/2.add-two-numbers.cpp
/*
 * @lc app=leetcode id=2 lang=cpp
 *
 * [2] Add Two Numbers
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0;
        int sum;

        ListNode dummy;

        ListNode* curr, * prev;
        prev = &dummy;

        while(l1 && l2) {
            sum = carry + l1->val + l2->val;

            carry = sum / 10;
            curr = new ListNode(sum % 10);

            prev->next = curr;
            prev = prev->next;

            l1 = l1->next;
            l2 = l2->next;
        }

        while(l1) {
            sum = carry + l1->val;

            carry = sum / 10;
            curr = new ListNode(sum % 10);

            prev->next = curr;
            prev = prev->next;

            l1 = l1->next;
        }

        while(l2) {
            sum = carry + l2->val;

            carry = sum / 10;
            curr = new ListNode(sum % 10);

            prev->next = curr;
            prev = prev->next;

            l2 = l2->next;
        }

        if(carry) {
            prev->next = new ListNode(carry);
        }

        return dummy.next;
    }
};
// @lc code=end

